File: telemetry-py/telemetry/_private/logging/mcap.py
"""MCAP logging writer for Foxglove Studio.

This module provides MCAP file logging for visualization in Foxglove Studio.
Logs are written to MCAP files using the Foxglove Log schema, allowing
time-series visualization and analysis.

Typical usage example:

    mcap_logger = MCAPLogger(
        mcap_writer=writer,
        service_name="my-service",
        service_version="1.0.0",
        service_environment="production"
    )
    mcap_logger.write_log("INFO", "User action", {"user_id": "123"})
"""

import time
from typing import Dict, Any, Optional


class MCAPLogger:
    """MCAP logging writer for Foxglove Studio.

    Wraps the UnifiedMcapWriter to provide a simple interface for writing
    log entries to MCAP files with service metadata.

    Attributes:
        mcap_writer: The UnifiedMcapWriter instance.
        service_name: Name of the service.
        service_version: Version of the service.
        service_environment: Deployment environment.
    """

    def __init__(
        self,
        mcap_writer,
        service_name: str,
        service_version: str,
        service_environment: str,
    ):
        """Initialize the MCAP logger.

        Args:
            mcap_writer: UnifiedMcapWriter instance for writing to MCAP file.
            service_name: Name of the service for metadata.
            service_version: Version of the service.
            service_environment: Deployment environment (e.g., "production").
        """
        self.mcap_writer = mcap_writer
        self.service_name = service_name
        self.service_version = service_version
        self.service_environment = service_environment

    def write_log(
        self, level: str, message: str, data: Optional[Dict[str, Any]] = None
    ):
        """Write a log entry to the MCAP file.

        Writes a log record to the MCAP file using the Foxglove Log schema.
        The log includes timestamp, level, message, and structured data.

        Args:
            level: Log level string (DEBUG, INFO, WARNING, ERROR, CRITICAL).
            message: Log message text.
            data: Optional dictionary of structured data to include.

        Note:
            Logs are only written if the MCAP writer is open. Closed writers
            are silently ignored to prevent errors during shutdown.
        """
        if self.mcap_writer and not self.mcap_writer.is_closed():
            self.mcap_writer.write_log(
                level=level,
                message=message,
                data=data or {},
                timestamp=time.time_ns(),
                name=self.service_name,
                file="",
                line=0,
                service_version=self.service_version,
                service_environment=self.service_environment,
            )

File: telemetry-go/internal/logging/mcap.go
package logging

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/the-protobuf-project/telemetry/telemetry-go/internal/foxglove"
	"github.com/the-protobuf-project/telemetry/telemetry-go/options"
)

// LogMcapWriter wraps the unified MCAP writer with Foxglove Log schema support
type LogMcapWriter struct {
	unifiedWriter      *foxglove.UnifiedMcapWriter
	channelID          uint16
	serviceName        string
	serviceVersion     string
	serviceEnvironment string
}

// FoxgloveLog represents a log message following the Foxglove Log schema
// https://github.com/foxglove/schemas/blob/main/schemas/jsonschema/Log.json
// Extended with service_version and service_environment fields
type FoxgloveLog struct {
	Timestamp          FoxgloveTimestamp      `json:"timestamp"`
	Level              int32                  `json:"level"`               // 0=UNKNOWN, 1=DEBUG, 2=INFO, 3=WARNING, 4=ERROR, 5=FATAL
	Message            string                 `json:"message"`             // Log message
	Name               string                 `json:"name"`                // Process or node name (service name with version and env)
	File               string                 `json:"file"`                // Filename
	Line               uint32                 `json:"line"`                // Line number
	Data               map[string]interface{} `json:"data,omitempty"`      // Additional structured data
	ServiceVersion     string                 `json:"service_version"`     // Service version (e.g., "1.0.0")
	ServiceEnvironment string                 `json:"service_environment"` // Service environment (e.g., "development", "production")
}

// FoxgloveTimestamp represents a timestamp in Foxglove format
type FoxgloveTimestamp struct {
	Sec  uint32 `json:"sec"`  // Seconds since epoch
	Nsec uint32 `json:"nsec"` // Nanoseconds (0-999999999)
}

// LogLevel constants matching Foxglove Log schema
const (
	LogLevelUnknown = 0 // Unknown log level
	LogLevelDebug   = 1 // Debug log level
	LogLevelInfo    = 2 // Info log level
	LogLevelWarning = 3 // Warning log level
	LogLevelError   = 4 // Error log level
	LogLevelFatal   = 5 // Fatal log level
)

// NewLogMcapWriter creates a new log writer using the unified MCAP writer
func NewLogMcapWriter(serviceOpts options.ServiceOptions, unifiedWriter *foxglove.UnifiedMcapWriter) (*LogMcapWriter, error) {
	// Format service name with version and environment
	serviceName := fmt.Sprintf("%s (%s | %s)", serviceOpts.Name, serviceOpts.Version, serviceOpts.Environment)

	// Topic for logs
	topic := fmt.Sprintf("/logs/%s", serviceOpts.Name)

	// Channel metadata
	metadata := map[string]string{
		"service":     serviceOpts.Name,
		"version":     serviceOpts.Version,
		"environment": string(serviceOpts.Environment),
		"description": serviceOpts.Description,
	}

	// Create log channel in unified writer
	channelID, err := unifiedWriter.CreateLogChannel(topic, metadata)
	if err != nil {
		return nil, fmt.Errorf("failed to create log channel: %w", err)
	}

	return &LogMcapWriter{
		unifiedWriter:      unifiedWriter,
		channelID:          channelID,
		serviceName:        serviceName,
		serviceVersion:     serviceOpts.Version,
		serviceEnvironment: string(serviceOpts.Environment),
	}, nil
}

// WriteLog writes a log message using the Foxglove Log schema
func (l *LogMcapWriter) WriteLog(level, message, file string, line uint32, data map[string]interface{}) error {
	now := time.Now()

	// Convert string level to Foxglove level integer
	levelInt := stringToFoxgloveLevel(level)

	// Create Foxglove Log message
	logMsg := FoxgloveLog{
		Timestamp: FoxgloveTimestamp{
			Sec:  uint32(now.Unix()),
			Nsec: uint32(now.Nanosecond()),
		},
		Level:              levelInt,
		Message:            message,
		Name:               l.serviceName,
		File:               file,
		Line:               line,
		Data:               data,
		ServiceVersion:     l.serviceVersion,
		ServiceEnvironment: l.serviceEnvironment,
	}

	// Serialize to JSON
	msgData, err := json.Marshal(logMsg)
	if err != nil {
		return fmt.Errorf("failed to marshal log message: %w", err)
	}

	nowNano := uint64(now.UnixNano())
	return l.unifiedWriter.WriteMessage(l.channelID, msgData, nowNano, nowNano)
}

// Close is a no-op since the unified writer is managed at the Telemetry level
func (l *LogMcapWriter) Close() error {
	return nil
}

// IsClosed returns whether the writer is closed
func (l *LogMcapWriter) IsClosed() bool {
	return l.unifiedWriter.IsClosed()
}

// GetFilePath returns the path to the MCAP file
func (l *LogMcapWriter) GetFilePath() string {
	return l.unifiedWriter.GetFilePath()
}

// stringToFoxgloveLevel converts string log level to Foxglove integer level
func stringToFoxgloveLevel(level string) int32 {
	switch level {
	case "DEBUG", "debug", "Debug":
		return LogLevelDebug
	case "INFO", "info", "Info":
		return LogLevelInfo
	case "WARN", "warn", "Warn", "WARNING", "warning", "Warning":
		return LogLevelWarning
	case "ERROR", "error", "Error":
		return LogLevelError
	case "FATAL", "fatal", "Fatal":
		return LogLevelFatal
	default:
		return LogLevelUnknown
	}
}
